flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/factories/TestSupportsStagingTableFactory.java

```diff
@@ -0,0 +1,171 @@
+        return Collections.emptySet();
+    }
+
+    /** A sink that supports staging. */
+    private static class SupportsStagingTableSink implements DynamicTableSink, SupportsStaging {
+
+        private String dataDir;
+        private TestStagedTable stagedTable;
+
+        public SupportsStagingTableSink(String dataDir) {
+            this(dataDir, null);
+        }
+
+        public SupportsStagingTableSink(String dataDir, TestStagedTable stagedTable) {
+            this.dataDir = dataDir;
+            this.stagedTable = stagedTable;
+        }
+
+        @Override
+        public ChangelogMode getChangelogMode(ChangelogMode requestedMode) {
+            return ChangelogMode.insertOnly();
+        }
+
+        @Override
+        public SinkRuntimeProvider getSinkRuntimeProvider(Context context) {
+            return new DataStreamSinkProvider() {
+                @Override
+                public DataStreamSink<?> consumeDataStream(
+                        ProviderContext providerContext, DataStream<RowData> dataStream) {
+                    if (stagedTable != null) {
+                        return dataStream
+                                .addSink(new StagedSinkFunction(dataDir))
+                                .setParallelism(1);
+                    } else {
+                        // otherwise, do nothing
+                        return dataStream.addSink(new DiscardingSink<>());
+                    }
+                }
+            };
+        }
+
+        @Override
+        public DynamicTableSink copy() {
+            return new SupportsStagingTableSink(dataDir, stagedTable);
+        }
+
+        @Override
+        public String asSummaryString() {
+            return "SupportsStagingTableSink";
+        }
+
+        @Override
+        public StagedTable applyStaging(StagingContext context) {
+            stagedTable = new TestStagedTable();
+            return stagedTable;
+        }
+    }
+
+    /** The sink for delete existing data. */
+    private static class StagedSinkFunction extends RichSinkFunction<RowData> {
+
+        private String dataDir;
+
+        public StagedSinkFunction(String dataDir) {
+            this.dataDir = dataDir;
+        }
+
+        @Override
+        public void open(Configuration parameters) throws Exception {
+            super.open(parameters);
+            File parentDir = new File(dataDir);
+            if (parentDir.exists()) {
+                parentDir.delete();
+            }
+            parentDir.mkdirs();
+            new File(dataDir, "data").createNewFile();
+        }
+
+        @Override
+        public void invoke(RowData value, Context context) throws Exception {
+            FileUtils.writeFileUtf8(
+                    new File(dataDir, "data"), value.getInt(0) + "," + value.getString(1));
+        }
+    }
+
+    /** A StagedTable for test. */
+    private static class TestStagedTable implements StagedTable {
+
+        @Override
+        public void begin() {}
```

Comment:
If we do nothing in these method, we can never know these method will be really called when job begin, finish, and abort.